### DISCUSSION TITLE
element property

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#1947] [Q&A] [answered]
in MicroStation ,i see the properties of the element  property group  like  "General","Material","Extentded" , and property like 'Level' ,'Range Low' and so on.
(in the \imodeljs-master\test-apps\presentation-test-app )  i use the google chrome  open the bim file (dgn file change by itwins snapshot), and show the properties of the element just  "user Label" and so on ,but not show properties like 'Level' ,'Range Low' .
so i want know : 1. if the properties like  'Level' ,'Range Low' not export in ibim file  by itwins snapshot, how to config  itwins snapshot to export? 2. if the properties like  'Level' ,'Range Low'   export in ibim file,please how show these properties in web by imodeljs .thank you!
 
![msproperties](https://user-images.githubusercontent.com/18022720/127111482-093e00ec-3f36-4e4f-bd11-b0083b6155fc.jpg)
![imodeljsproperties](https://user-images.githubusercontent.com/18022720/127111497-ed4b7190-d4e6-4dc4-a5f4-9c6b4491704a.jpg)

**Comment by pmconne** [comment]
Hi,

Levels are converted to [categories](https://www.itwinjs.org/bis/intro/categories/) (or sometimes subcategories).

Most of the other properties shown in your MicroStation screenshot have to do with the element's geometry and/or symbology, and many are computed by MicroStation, not stored directly.
iModel.js is based on [BIS](https://www.itwinjs.org/bis/) which is concerned with modeling real-world objects like pumps, walls, pipelines, etc, as opposed to "dumb" abstract objects like spheres, slabs, and cylinders. The [geometry stream](https://www.itwinjs.org/learning/common/geometrystream/) representing such real-world objects tends to consist of multiple types of geometry and different symbologies, so talking about the object in terms of its raw geometry is less useful than talking about its real-world engineering properties.

You can obtain the range, origin, and rotation of one or more [geometric elements](https://www.itwinjs.org/bis/domains/biscore.ecschema/#geometricelement) using [IModelConnection.elements.getPlacements](https://github.com/imodeljs/imodeljs/blob/3b9050e4a62e52357e3489264fefe711af701b2d/core/frontend/src/IModelConnection.ts#L914).
Properties like area, volume, and perimeter can be computed using [IModelConnection.getMassProperties](https://www.itwinjs.org/reference/imodeljs-frontend/imodelconnection/imodelconnection/).
You can iterate the individual geometric primitives and symbology changes stored in an element's geometry stream using [GeometryStreamIterator](https://www.itwinjs.org/reference/imodeljs-common/geometry/geometrystreamiterator/).

**Comment by grigasp** [comment]
I would also add that some of the geometry-related properties are persisted in the iModel, but hidden using custom attributes in the BIS schema. 

For GeometricElement2d: Origin, Rotation, BBoxLow, BBoxHigh.
For GeometricElement3d: Origin, Yaw, Pitch, Roll, BBoxLow, BBoxHigh.

One way to override that custom attribute is to add a `ContentModifier` presentation rule like this:
```json
{
  "ruleType": "ContentModifier",
  "class": { "schemaName": "BisCore", "className": "GeometricElement3d" },
  "propertyCategories": [
    {
      "id": "geometric",
      "label": "Geometric Properties"
    }
  ],
  "propertyOverrides": [
    {
      "name": "Yaw",
      "categoryId": "geometric",
      "isDisplayed": true,
      "doNotHideOtherPropertiesOnDisplayOverride": true
    },
    {
      "name": "Pitch",
      "categoryId": "geometric",
      "isDisplayed": true,
      "doNotHideOtherPropertiesOnDisplayOverride": true
    },
    {
      "name": "Roll",
      "categoryId": "geometric",
      "isDisplayed": true,
      "doNotHideOtherPropertiesOnDisplayOverride": true
    },
    {
      "name": "Origin",
      "categoryId": "geometric",
      "isDisplayed": true,
      "doNotHideOtherPropertiesOnDisplayOverride": true
    },
    {
      "name": "BBoxLow",
      "categoryId": "geometric",
      "isDisplayed": true,
      "doNotHideOtherPropertiesOnDisplayOverride": true
    },
    {
      "name": "BBoxHigh",
      "categoryId": "geometric",
      "isDisplayed": true,
      "doNotHideOtherPropertiesOnDisplayOverride": true
    }
  ]
}
```

It generates an additional group in the property grid:
![image](https://user-images.githubusercontent.com/35135765/127119688-fb6f051c-1bba-4b19-9627-0eed184a8c8c.png)

**Reply by pmconne** [reply]
Note that "bounding box" is not the same as "range" - the former is relative to the element's local coordinate system, the latter to the model's coordinate system.

**Reply by paprophet** [reply]
> Note that "bounding box" is not the same as "range" - the former is relative to the element's local coordinate system, the latter to the model's coordinate system.

thank you, how can get the General-Transparency and  Material-Attached Material properties ?

**Reply by pmconne** [reply]
That is a complicated question because there is no 1:1 mapping between those MicroStation properties and any properties of an iModel element.
Each GeometricElement belongs to one Category ("level"). Categories have one or more SubCategories, including a default SubCatetgory.
Each SubCategory has a SubCategoryAppearance (similar to level appearance in MicroStation) specifying its default symbology.
Each entry in the GeometryStream can specify the Id of a subcategory to use for the base symbology; if none is specified the default subcategory is used.
Each entry can then optionally override some or all of the subcategory's symbology.

There is not currently a convenient frontend API to resolve the symbology.
You can approximate it by:
- Obtain the element's properties, including its geometry stream, via `IModelConnection.elements.loadElement(elementId, { wantGeometry: true })`.
- Iterate the geometry stream using a GeometryStreamIterator. Some details: https://www.itwinjs.org/learning/common/geometrystream/#iterating-a-geometrystream
- For each entry, inspect the GeometryParams
  - If a subcategoryId is specified, query its appearance from the backend
  - Check if `materialId` or `elmTransparency` is defined on the GeometryParams; if so they override that specified by the subcategory.
  - 
@bbastings anything to add? Any relevant documentation articles or APIs?

**Comment by pmconne** [comment, marked as answer]
I will assume lack of further discussion indicates answers provided are sufficient. If not, please provide further details.
